File: src/main/java/gargoyle/ct/prop/impl/CTBaseProperty.java
package gargoyle.ct.prop.impl;

import gargoyle.ct.prop.CTProperty;

import java.text.MessageFormat;

public abstract class CTBaseProperty<T> implements CTProperty<T> {

    protected final String   name;
    private final   Class<T> type;

    protected CTBaseProperty(Class<T> type, String name) {
        this.type = type;
        this.name = name;
    }

    public T get(T def) {
        T value = get();
        return value == null ? def : value;
    }

    @Override
    public String name() {
        return name;
    }

    @Override
    public Class<T> type() {
        return type;
    }

    @Override
    public String toString() {
        return MessageFormat.format("CTBaseProperty'{'name=''{0}'''}'", name);
    }
}

File: src/main/kotlin/gargoyle/ct/prop/impl/CTBaseProperty.kt
package gargoyle.ct.prop.impl

import gargoyle.ct.prop.CTProperty
import java.text.MessageFormat
import kotlin.reflect.KClass

abstract class CTBaseProperty<T : Any> protected constructor(type: KClass<T>, name: String) : CTProperty<T> {
    protected val name: String
    private val type: KClass<T>

    init {
        this.type = type
        this.name = name
    }

    open operator fun get(def: T): T {
        return get()
    }

    override val isPresent: Boolean
        get() = true

    override fun name(): String {
        return name
    }

    override fun type(): KClass<T> {
        return type
    }

    override fun toString(): String {
        return MessageFormat.format("CTBaseProperty'{'name=''{0}'''}'", name)
    }
}
